### PR TITLE
[7947] Feature: Blank out app preview screen when switching apps on m…

### LINKED ISSUE #7947: Feature: Blank out app preview screen when switching apps on mobile
# Feature Request

As a user, I'd like the app preview screen to be blanked out when switching apps.  This would increase my privacy.

As a bonus, I would like this to be a user option, but on by default.

### PR DESCRIPTION
**Pull request by bitsikka**
Fixes #7947 

Notes: https://github.com/status-im/status-react/pull/8009#issuecomment-490671670

status: ready

### CODE REVIEW

**Comment by bitsikka** [comment]
blanked out screen looks like this

image

**Comment by bitsikka** [comment]
screen with Share panel looks like this

image

**Comment by bitsikka** [comment]
Screen that has an Alert box active looks like this

image

**Comment by bitsikka** [comment]
Tested only in iOS simulator so far

File: src/status_im/ui/screens/views.cljs [hunk 1]
```diff
@@ -35,6 +37,19 @@
...
       [bottom-sheet/bottom-sheet opts])))
 
+(views/defview inactive-overlay []
+  (letsubs [app-state [:get :app-state]]
+    {:component-will-mount
+     (fn []
+       (react/dismiss-keyboard!))
+     :component-will-update
+     (fn []
+       (react/dismiss-keyboard!))}
```
Comment (flexsurfer) [inline src/status_im/ui/screens/views.cljs]: just wondering why do we need to dismiss keyboard on an update?

**Review by flexsurfer — APPROVED** [review]

File: src/status_im/ui/screens/views.cljs (hunk 1 above)
Comment (bitsikka) [reply src/status_im/ui/screens/views.cljs]: oh :) I was looking for all the cases where some UI elements would render above the blanked out component and see if I could deal with to not make it so. Keyboard was the only one I could deal with like that :) in that hacky way. Couldn't deal with `alert` and `share` as you can see

should I omit it?

I should probably also not use :get subscription considering `re-frame subscriptions optimisation`

**Comment by yenda** [comment]
Have we considered this approach? https://stackoverflow.com/questions/9822076/how-do-i-prevent-android-taking-a-screenshot-when-my-app-goes-to-the-background

**Comment by bitsikka** [comment]
> Have we considered this approach? https://stackoverflow.com/questions/9822076/how-do-i-prevent-android-taking-a-screenshot-when-my-app-goes-to-the-background

didn't know about this.. sounds more robust
will check it out

**Comment by rasom** [comment]
yep seems that on android we need to try `FLAG_SECURE`

https://developer.android.com/reference/android/view/Display.html#FLAG_SECURE
https://github.com/kristiansorens/react-native-flag-secure-android

**Comment by bitsikka** [comment]
> yep seems that on android we need to try `FLAG_SECURE`
> 
> https://developer.android.com/reference/android/view/Display.html#FLAG_SECURE
> https://github.com/kristiansorens/react-native-flag-secure-android

Thanks @rasom @yenda for the pointers

I barely got my development machine in order after having it being broken more than a week ago.

I am hardly familiar with lower React-Native / Android development / Java areas, but this will give me a chance to get to know it

Going to do the due diligence and make a proper implementation soon

Please bare with me

**Comment by bitsikka** [comment]
@flexsurfer @yenda @rasom 

Changed the implementation as guided using FLAG_SECURE in android. Additionally, used SharedPreferences/NSUserDefaults to persist flag setting - additions are made to Status native-module.

@corpetty includes bonus (user setting) as requested

Tested in iOS Simulator and Android-avd:

- tested preview is blank by default
- tested preview is blank/not-blank according to user setting
- tested preview setting persists through app re-starts

### Notes:
- in Android (when preview privacy mode is on) prevents screen capture in non preview screen as well
- in iOS does not prevent screen capture
- in iOS if there is keyboard showing at the time of putting app to background, keyboard renders in preview above blank screen

**Comment by flexsurfer** [comment]
@rasom @yenda ping

**Comment by flexsurfer** [comment]
@mandrigin @rasom @yenda ping

File: ios/StatusIm/AppDelegate.m [hunk 2]
```diff
@@ -101,4 +105,36 @@ - (NSURL *)sourceURLForBridge:(RCTBridge *)bridge
 #endif
 }
 
+- (void)applicationWillResignActive:(UIApplication *)application {
+  if ([[NSUserDefaults standardUserDefaults] boolForKey:@"BLANK_PREVIEW"]) {
+    // fill screen with our own colour
+    UIView *colourView = [[UIView alloc]initWithFrame:self.window.frame];
```
Comment (mandrigin) [inline ios/StatusIm/AppDelegate.m]: I think it is better to just save it as a local variable than a view with a tag.

File: android/app/src/main/java/im/status/ethereum/MainActivity.java [hunk 3]
```diff
@@ -202,6 +203,18 @@ private void rejectRootedNotification() {
         editor.commit();
     }
 
+    private static final String BLANK_PREVIEW = "BLANK_PREVIEW";
```
Comment (mandrigin) [inline android/app/src/main/java/im/status/ethereum/MainActivity.java]: hmm, the same constant declared twice kind of defeat the purpose of having a constant in the first place. Maybe we can just make a static class like `Constants` or `Preferences.BLANK_PREVIEW` and use it in both places.

**Review by mandrigin — CHANGES_REQUESTED** [review]
a few small changes, good job otherwise!

**Comment by StatusSceptre** [comment]
@flexsurfer @mandrigin - can we consider this PR complete so I can release the funds for the bounty?

File: android/app/src/main/java/im/status/ethereum/MainActivity.java (hunk 3 above)
Comment (bitsikka) [reply android/app/src/main/java/im/status/ethereum/MainActivity.java]: I think we can do without the constant. Changed the implementation a little bit in update that is coming shortly.

File: ios/StatusIm/AppDelegate.m (hunk 2 above)
Comment (bitsikka) [reply ios/StatusIm/AppDelegate.m]: done

**Comment by bitsikka** [comment]
@mandrigin updated PR with suggested changes.

Please note that at this point I am barely getting acquainted with Java and Objective-c and the implementation might still be sloppy. Hope the work is acceptable. If not, with feedback, still going to do what it takes.

Those (Java/Obj-c) parts have mostly been an adaptation(with prior knowledge of languages other than those) of code found in the interwebs; with finally reading documentations in the past 24 hrs :p.  Sorry about that and thanks for baring with me.

Also definitely going to continue to spend some time acquiring knowledge on them for knowledge sake.

**Review by mandrigin — APPROVED** [review]

**Comment by bitsikka** [comment]
> An unknown server-side error occurred while processing the command. Original error: io.appium.uiautomator2.common.exceptions.TakeScreenshotException: Failed to capture a screenshot. Does the current view have 'secure' flag set? at io.appium.uiautoma

yes 'secure' flag is set

@antdanchenko @churik help! looks like e2e tests are going to have to be updated for this one as well

or the implementation is going to have to change if significant number of tests depend on capturing screenshot 🤔

**Comment by mandrigin** [comment]
@bitsikka maybe we can have this option behind a build flag somehow? so it is disabled for e2e?

**Comment by mandrigin** [comment]
@bitsikka or, easier, it should be `false` by default (it is `true` for Android in your code).

**Comment by bitsikka** [comment]
> @bitsikka or, easier, it should be `false` by default (it is `true` for Android in your code).

yes, it is `true` by default for both Android and iOS(for iOS, just the blanking out in background state - not preventing screencap). @corpetty OP of the issue, requested it that it be so, and it made sense to do that, but we had no idea of this consequence

I'll make it `false` cause now, that also makes sense :)

**Comment by churik** [comment]
Thanks, @bitsikka!

Tested IOS 11.4.1 (IPhone 7), Android 8 (LG V20):
- going to background with wallet, chat screens, with popup, with opened keyboard, on login screen - preview is blanked out;
- reopening app;

**Review by churik — APPROVED** [review]

**Pull request merged**